embassy-sync/src/mutex.rs

```diff
@@ -17,8 +17,39 @@ use crate::waitqueue::WakerRegistration;
 pub struct TryLockError;
 
 struct State {
-    locked: bool,
-    waker: WakerRegistration,
+    locked: Cell<bool>,
+    waker: UnsafeCell<WakerRegistration>,
+}
+
+impl State {
+    const fn new() -> Self {
+        Self {
+            locked: Cell::new(false),
+            waker: UnsafeCell::new(WakerRegistration::new()),
+        }
+    }
+
+    fn lock(&self, waker: &core::task::Waker) -> bool {
+        if self.locked.replace(true) {
+            unsafe { (&mut *self.waker.get()).register(waker) };
```

Comment:
I don't think this is sound. `register()` can drop or wake wakers, which will call waker methods through the vtable which is basically running arbitrary code. This code could call back into the same Mutex, causing two `&mut`s to the same WakerRegistration to exist at the same time, which is UB.

Reply:
it might be possible if it uses `Cell<Option<Waker>>`, but it'd mean you'd have to constantly take out and put back the waker so it might be less efficient, not sure.

Reply:
We can look into it through a pointer I think.

Reply:
I don't think so. for example to wake the waker you'd need a `&Waker` pointing to the contents. The wake can run arbitrary code which can reentrantly call into the same WakerRegistration and replace the waker (with either `&mut` or `ptr::write`), which would be UB because you'd be mutating the data while a `&Waker` pointing to that location exists.

Reply:
Uh, I might be dense here to not understand, but given the following impl, how can a reference exist while we're mutating the cell?

```rust
fn register(&mut self, w: &Waker) {
        match self.inner_as_ref() {
            Some(ref w2) if (w2.will_wake(w)) => return,
            _ => {}
        }

        // clone the new waker and store it
        if let Some(old_waker) = self.waker.replace(Some(w.clone())) {
            old_waker.wake() // this may call reentrantly into `register`
        }
    }
```